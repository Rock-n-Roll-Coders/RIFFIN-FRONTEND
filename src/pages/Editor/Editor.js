import { useState, useContext, useEffect } from 'react'
import { useNavigate, useParams } from "react-router-dom";
import { UserContext } from "../../App";
import * as tablatureServices from "../../services/tablatureServices";
import * as userUtils from "../../utils/userUtils";
import { CircularProgress } from "@mui/material";
import BarGroup from './components/BarGroup';
import Controls from './components/Controls'

import Box from '@mui/material/Box';

const Editor = () => {
  const [showDeleteButton, setShowDeleteButton] = useState(false); // Is the document already in the database?
  // TODO Rename: isWaitingForResponse
  const [isLoading, setIsLoading] = useState(false); // is the document currently waiting for a response?
  const [tablature, setTablature] = useState({
    isPublic: false,
    name: "A tasty lick",
    bars: [],
    tags: [],
    isBassTab: false,
  });
   const { user } = useContext(UserContext);
  const { tabId } = useParams();
  let navigate = useNavigate();
  
  const deleteTablatureFromDatabase = () => {
    const idToken = userUtils.getIdTokenFromUser(user);
    tablatureServices.delete(tablature._id, idToken).then((res) => {
      console.log(res);
    });
    // TODO Navigate to trending
  };

  const saveTablatureToDatabase = () => {
    const updateExistingTablature = (idToken) => {
      setIsLoading(true);
      tablatureServices.update(tablature, idToken).then((res) => {
        console.log(res);
        setIsLoading(false);
        setTablature({ ...tablature });
      });
    }
  
    const saveNewTablature = (idToken) => {
      setIsLoading(true);
      tablatureServices
        .create(tablature, idToken)
        .then((tablatureFromResponse) => {
          setIsLoading(false);
          navigate(`/edit/${tablatureFromResponse._id}`);
        });
    }

    const idToken = userUtils.getIdTokenFromUser(user);
    if (tablature._id) {
      updateExistingTablature(idToken);
    } else {
      saveNewTablature(idToken);
    }
  };

  const deleteBarFromTablature = (barIndex) => {
    // TODO Find a way for this be done through BarController?
    const newBars = [];
    tablature.bars.forEach((bar, i) => {
      if (barIndex !== i) {
        newBars.push({ ...bar });
      }
    });

    tablature.bars = newBars;
    setTablature({ ...tablature });
  };

  const addBarToTablature = () => {
    // TODO Find a way for this be done through BarController?
    const mapOfLastColumnIndexes = {
      40: true,
      81: true,
      122: true,
      163: true,
      204: true,
      245: true,
    };

    const initTextAreaWithValue = (character) => {
      let charactersInString = [];
      for (let i = 0; i < 245; i++) {
        if (i in mapOfLastColumnIndexes) {
          charactersInString.push("\n");
        } else {
          charactersInString.push(character);
        }
      }
      return charactersInString.join("");
    };

    const previousBars = [];
    tablature.bars.forEach((bar) => {
      previousBars.push({ ...bar });
    });

    const newBar = {
      label: `Bar ${tablature.bars.length + 1}`,
      tempKey: Date() + Math.random(),
      inputs: initTextAreaWithValue(" "),
      dashes: initTextAreaWithValue("-"),
    };

    tablature.bars = [...previousBars, newBar];
    setTablature({ ...tablature });
  };

  const setPublic = () => {
    const udpatedTablature = {
      ...tablature,
      isPublic: !tablature.isPublic,
    };
    setTablature(udpatedTablature);
  };

  const handleNameInput = (event) => {
    const udpatedTablature = {
      ...tablature,
      name: event.target.value,
    };
    setTablature(udpatedTablature);
  };

  const refreshTablatureObject = () => {
    setTablature({ ...tablature });
  }

  // Check if the document is new
  useEffect(() => {
    tablature["_id"] ? setShowDeleteButton(true) : setShowDeleteButton(false);
  }, [tablature]);

  useEffect(() => {
    if (tabId) {
      setIsLoading(true)
      tablatureServices.getTablatureById(tabId).then((res) => {
        if (res["error"]) {
          // TODO Navigate back to where the user came from
          navigate(`/trending`);
        }
        setTablature(res.tablature);
        setIsLoading(false)
      });
    }
  }, [tabId]);
  
  useEffect(() => {
    if (user) {
      let username = user.username;
      setTablature((prev) => {
        return { ...prev, owner: username };
      });
    }
  }, [user]);

  useEffect(() => {
    if(!tabId && tablature.bars.length === 0) {
      addBarToTablature()
    }
  }, [tablature, tabId])

  return (
    <>
      {isLoading ? ( <CircularProgress /> ) : (
        <>
          <Box sx={{display: 'flex', justifyContent: 'space-between'}}>
            <input
              type="text"
              name="name"
              value={tablature.name}
              onChange={handleNameInput}
              placeholder="A tasty lick"
            />

            <Controls 
              deleteTablatureFromDatabase={deleteTablatureFromDatabase}
              allowDelete={showDeleteButton}
              setPublic={setPublic}
              isPublic={tablature.isPublic}
              saveTablatureToDatabase={saveTablatureToDatabase}
              addBarToTablature={addBarToTablature}    
            />
          </Box>

          <BarGroup
            bars={tablature.bars} 
            refreshTablatureObject={refreshTablatureObject}
            deleteBarFromTablature={deleteBarFromTablature}
          />
        </>
      )}
    </>
  );
}
 
export default Editor;