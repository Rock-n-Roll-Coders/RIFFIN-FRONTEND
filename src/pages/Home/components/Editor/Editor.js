// Components / hooks
import { useState, useContext, useEffect } from 'react';
import { useNavigate, useParams } from "react-router-dom";
import { UserContext } from "containers/CognitoUserProvider/CognitoUserProvider";
import { TablatureContext } from "containers/TablatureProvider/TablatureProvider";
import ExpandableBar from './components/ExpandableBar/ExpandableBar';
import AddBlockButton from './components/AddBlockButton/AddBlockButton';
import DeleteTabButton from './components/DeleteTabButton/DeleteTabButton';
import SaveTabButton from './components/SaveTabButton/SaveTabButton';

// Services / utils
import { getNewGuitarBlock } from "./utils/EditorUtils";

// MUI
import { CircularProgress, Paper } from "@mui/material";
import Box from '@mui/material/Box';

const Editor = (props) => {
  const [selectedBar, setSelectedBar] = useState(null);
  const [cursorPosition, setCursorPosition] = useState({ position: null }); // This can't be a number or it will cause 
  const [showDeleteButton, setShowDeleteButton] = useState(false); // Is the document already in the database?
  const [isLoading, setIsLoading] = useState(false); // is the document currently waiting for a response?
  const [tablature, setTablature] = useState({
    isPublic: false,
    name: "A tasty lick",
    bars: [],
    tags: [],
    isBassTab: false,
  });

  const { user } = useContext(UserContext);
  const { tabId } = useParams();
  const { getTabFromUser } = useContext(TablatureContext)
  let navigate = useNavigate();

  const toggleLoading = (value) => setIsLoading(value);
  
  function handleAddCharacter(character, mapOfLastColumnIndexes) {
    if (cursorPosition.position in mapOfLastColumnIndexes) {
      setCursorPosition((prev) => {
        return { position: prev.position };
      });
    } else {
      const bar = tablature.bars[selectedBar.index];
      const newBar = { ...bar };
      newBar.inputs = getUpdatedTextAreaValues(
        "inputs",
        character,
        cursorPosition.position
      );
      newBar.dashes = getUpdatedTextAreaValues(
        "dashes",
        " ",
        cursorPosition.position
      );
      tablature.bars[selectedBar.index] = newBar;
      setTablature((prev) => { return {...prev}})
      setCursorPosition({ position: cursorPosition.position + 1 });
    }
  }

  function handleRemoveCharacter(mapOfFirstColumnIndexes) {
    if (cursorPosition.position in mapOfFirstColumnIndexes) {
      setCursorPosition((prev) => {
        return { position: prev.position };
      });
    } else {
      const bar = tablature.bars[selectedBar.index];
      const newBar = { ...bar };
      let newCursorPosition = cursorPosition.position - 1;
      newBar.inputs = getUpdatedTextAreaValues(
        "inputs",
        " ",
        newCursorPosition
      );
      newBar.dashes = getUpdatedTextAreaValues(
        "dashes",
        "-",
        newCursorPosition
      );
      tablature.bars[selectedBar.index] = newBar;
      setTablature((prev) => { return {...prev}})
      setCursorPosition({ position: newCursorPosition });
    }
  }

  const legalCharacters = {
    "~": handleAddCharacter, // vibrato
    "/": handleAddCharacter, // slide
    "^": handleAddCharacter, // bend
    "x": handleAddCharacter, // mute
    "p": handleAddCharacter, // pull off
    "h": handleAddCharacter, // hammer on
    0: handleAddCharacter,
    1: handleAddCharacter,
    2: handleAddCharacter,
    3: handleAddCharacter,
    4: handleAddCharacter,
    5: handleAddCharacter,
    6: handleAddCharacter,
    7: handleAddCharacter,
    8: handleAddCharacter,
    9: handleAddCharacter,
    "d": handleAddCharacter, // duplicate
    "Backspace": handleRemoveCharacter,
  //   insertLineBreak: handlePressingEnter, // Move cursor to the next line (string)
  };

  function getUpdatedTextAreaValues(area, character, insertionIndex) {
    const bar = tablature.bars[selectedBar.index];
    let currentValueAsArray = [...bar[area]];
    currentValueAsArray[insertionIndex] = character;
    const updatedValue = currentValueAsArray.join("");
    return updatedValue;
  }

  const arrows = {
    ArrowDown: true,
    ArrowLeft: true,
    ArrowRight: true,
    ArrowUp: true,
  };

  const deleteBarFromTablature = (barIndex) => {
    const newBars = [];
    tablature.bars.forEach((bar, i) => {
      if (barIndex !== i) {
        newBars.push({ ...bar });
      }
    });

    tablature.bars = newBars;
    setTablature({ ...tablature });
  };

  const duplicateBlock = (blockIndex) => {
    const newBlock = {
      tempKey: Date() + Math.random(),
      label: tablature.bars[blockIndex].label,
      inputs: tablature.bars[blockIndex].inputs,
      dashes: tablature.bars[blockIndex].dashes,
      cols: tablature.bars[blockIndex].cols,
      maxLength: tablature.bars[blockIndex].maxLength
    }
    tablature.bars.push(newBlock)
    refreshTablatureObject()
  }

  const handleNameInput = (event) => {
    const udpatedTablature = {
      ...tablature,
      name: event.target.value,
    };
    setTablature(udpatedTablature);
  };

  const refreshTablatureObject = () => setTablature({ ...tablature });

  const handleClickedBar = (event, barIndex, barRef) => {
    setSelectedBar({ inputRef: barRef, index: barIndex });
    setCursorPosition({ position: event.target.selectionStart });
  };

  const handleKeyUpInBar = (event) => {
    if (event.key in arrows) {
      setCursorPosition({ position: event.target.selectionStart });
    }
  };

  const handleBarChange = (event, mapOfLastColumnIndexes, mapOfFirstColumnIndexes) => {
    const key = event.nativeEvent.data;
    console.log(event.nativeEvent.inputType);
    event.preventDefault();
    if (key in legalCharacters) {
      console.log(key);
      legalCharacters[key](key, mapOfLastColumnIndexes);
    } else if (key === null) {
      if (event.nativeEvent.inputType === "deleteContentBackward") {
        legalCharacters["Backspace"](mapOfFirstColumnIndexes);
      } else if (event.nativeEvent.inputType === "insertLineBreak") {
        legalCharacters["insertLineBreak"]();
      }
    } else {
      setCursorPosition((prev) => {
        return { position: prev.position };
      });
    }
  };

  // Prevent crusor from jumping to end of input.
  useEffect(() => {
    if (selectedBar) {
      selectedBar.inputRef.current.selectionStart = cursorPosition.position;
      selectedBar.inputRef.current.selectionEnd = cursorPosition.position;
      console.log("New cursorPosition: ", cursorPosition.position);
    }
  }, [cursorPosition, selectedBar]);

  // Check if the document is new
  useEffect(() => {
    tablature["_id"] ? setShowDeleteButton(true) : setShowDeleteButton(false);
  }, [tablature]);

  useEffect(() => {
    if (tabId) {
      const tab = getTabFromUser(tabId)
      if(tab) {
        setTablature(getTabFromUser(tabId));
      } else {
        navigate('/trending')
      }
    }
  }, [tabId, getTabFromUser, navigate]);
  
  useEffect(() => {
    if (user) {
      let username = user.username;
      setTablature((prev) => {
        return { ...prev, owner: username };
      });
    }
  }, [user]);

  useEffect(() => {
    if(!tabId && tablature.bars.length === 0) {
      const newBlock = getNewGuitarBlock()
      setTablature((prev) => {
        prev.bars = [newBlock]
        return {...prev}
      })
    }
  }, [tabId, tablature.bars.length])

  return (
    <div data-testid="Editor">
      {isLoading ? ( <CircularProgress /> ) : (
        <Paper>
          <Box sx={{display: 'flex', justifyContent: 'space-between'}}>
            <input
              type="text"
              name="name"
              value={tablature.name}
              onChange={handleNameInput}
              placeholder="A tasty lick"
            />
            <AddBlockButton 
              tablature={tablature}
              refreshTablatureObject={refreshTablatureObject}
            />
            <SaveTabButton 
              tablature={tablature}
              toggleLoading={toggleLoading}
              tags={props.tags}
              refreshTablatureObject={refreshTablatureObject}
            />
            {showDeleteButton &&
              <DeleteTabButton 
                tablature_id={tablature._id}
              />
            }
          </Box>
          {tablature.bars.map((bar, i) => (
            <ExpandableBar
              key={i}
              index={i}
              bar={bar}
              duplicateBlock={duplicateBlock}
              deleteBarFromTablature={deleteBarFromTablature}
              handleBarChange={handleBarChange}
              handleKeyUpInBar={handleKeyUpInBar} 
              handleClickedBar={handleClickedBar}
              refreshTablatureObject={refreshTablatureObject}
            />
          ))}
        </Paper>
      )}
    </div>
  );
}
 
export default Editor;