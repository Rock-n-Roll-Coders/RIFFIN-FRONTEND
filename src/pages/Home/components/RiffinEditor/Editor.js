// Components / hooks
import AddNewBlockButton from "./components/AddNewBlockButton/AddNewBlockButton";
import TitleInput from "./components/TitleInput/TitleInput";
import DeleteTabButton from "./components/DeleteTabButton/DeleteTabButton";
import SaveTabButton from "./components/SaveTabButton/SaveTabButton";
import TablatureBlock from "./components/TablatureBlock/TablatureBlock";
import { useContext } from "react";
import { RiffinEditorDispatch } from "./RiffinProvider";
// MUI
import { Box, Grid } from "@mui/material";
const Content = () => {
  const { editor } = useContext(RiffinEditorDispatch);
  return (
    <>
        <Grid container rowSpacing={2} columnSpacing={4} sx={{alignItems: "end"}}>
          <Grid item>
            <TitleInput />
          </Grid>
          <Grid item>
            <SaveTabButton />
          </Grid>
          <Grid item>
            <DeleteTabButton />
          </Grid>
        </Grid>
        {editor.tablature.blocks.map((block, i) => (
          <Box sx={{my: 4}} key={i}>
            <TablatureBlock key={i} index={i} block={block} />
          </Box>
        ))}
        <AddNewBlockButton />
    </>
  );
}
 
export default Content;