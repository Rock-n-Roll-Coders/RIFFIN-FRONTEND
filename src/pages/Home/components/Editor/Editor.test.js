import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { BrowserRouter } from "react-router-dom";
import Editor from "./Editor";

test('A tasty lick label loads on initialization', ()=> {
  render(<Editor />, {wrapper: BrowserRouter})
  expect(screen.getByPlaceholderText("A tasty lick")).toBeInTheDocument();
})

test('A bar should load on initialization', () => {
  render(<Editor />, {wrapper: BrowserRouter})
  const length = screen.getAllByPlaceholderText("a new bar").length
  expect(length).toBe(1)
})

test('Controls load on initialization', ()=> {
  render(<Editor />, {wrapper: BrowserRouter})
  expect(screen.getByLabelText(/save/i)).toBeInTheDocument()
  expect(screen.getByLabelText(/new bar/i)).toBeInTheDocument()
  expect(screen.getByLabelText(/make private/i)).toBeInTheDocument()
})

test('Clicking new bar button adds a bar to the bargroup', async () => {
  render(<Editor />, {wrapper: BrowserRouter})
  const user = userEvent.setup()
  
  for(let i = 1; i < 5; i++) {
    await user.click(screen.getByLabelText(/new bar/i))
    const length = screen.getAllByPlaceholderText("a new bar").length
    expect(length).toBe(i + 1)
  }

})

test('user can toggle between public and private', async () => {
  render(<Editor />, {wrapper: BrowserRouter})
  const user = userEvent.setup()
  expect(screen.getByLabelText(/make private/i)).toBeInTheDocument()
  await user.click(screen.getByLabelText(/make private/i))
  expect(screen.getByLabelText(/make public/i)).toBeInTheDocument()
  expect(screen.queryByLabelText(/make private/i)).not.toBeInTheDocument()
  await user.click(screen.getByLabelText(/make public/i))
  expect(screen.getByLabelText(/make private/i)).toBeInTheDocument()
  expect(screen.queryByLabelText(/make public/i)).not.toBeInTheDocument()
})

test('delete a bar removes a bar from the bars array', async () => {
  render(<Editor />, {wrapper: BrowserRouter})
  const user = userEvent.setup()
  
  for(let i = 1; i < 5; i++) {
    await user.click(screen.getByLabelText(/new bar/i))
    const length = screen.getAllByPlaceholderText("a new bar").length
    expect(length).toBe(i + 1)
  }

  await user.click(screen.getAllByLabelText(/delete bar/i))[0]
  const length = screen.getAllByPlaceholderText("a new bar").length
  expect(length).toBe(5)
})