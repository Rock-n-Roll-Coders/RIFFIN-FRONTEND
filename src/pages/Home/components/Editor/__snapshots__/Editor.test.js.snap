// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendersa new tab correctly 1`] = `
<div
  data-testid="Editor"
>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1ps6pg7-MuiPaper-root"
  >
    <div
      className="MuiBox-root css-gg4vpm"
    >
      <input
        name="name"
        onChange={[Function]}
        placeholder="A tasty lick"
        type="text"
        value="A tasty lick"
      />
      <button
        aria-label="Add tablature block"
        aria-labelledby={null}
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element={true}
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="PlaylistAddIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M14 10H3v2h11v-2zm0-4H3v2h11V6zm4 8v-4h-2v4h-4v2h4v4h2v-4h4v-2h-4zM3 16h7v-2H3v2z"
          />
        </svg>
        <span
          className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        aria-label="Add note block"
        aria-labelledby={null}
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element={true}
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="TextSnippetIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="m20.41 8.41-4.83-4.83c-.37-.37-.88-.58-1.41-.58H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V9.83c0-.53-.21-1.04-.59-1.42zM7 7h7v2H7V7zm10 10H7v-2h10v2zm0-4H7v-2h10v2z"
          />
        </svg>
        <span
          className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        aria-label="Save"
        aria-labelledby={null}
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element={true}
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="SaveIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7l-4-4zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3zm3-10H5V5h10v4z"
          />
        </svg>
        <span
          className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      className="MuiBox-root css-gg4vpm"
    >
      <input
        name="name"
        onChange={[Function]}
        placeholder="label"
        style={
          Object {
            "marginLeft": "10px",
          }
        }
        type="text"
      />
      <div>
        <button
          aria-haspopup="true"
          className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-disableElevation MuiButtonBase-root  css-zddlty-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          id="demo-customized-button"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          Options
          <span
            className="MuiButton-endIcon MuiButton-iconSizeMedium css-9tj150-MuiButton-endIcon"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="KeyboardArrowDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
              />
            </svg>
          </span>
          <span
            className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      className="MuiBox-root css-k008qs"
    >
      <textarea
        cols={1}
        readOnly={true}
        style={
          Object {
            "background": "transparent",
            "border": "none",
            "color": "#1976d2",
            "fontSize": "1.2rem",
            "margin": 0,
            "outline": "none",
            "padding": 0,
            "position": "relative",
            "resize": "none",
            "textAlign": "right",
            "zIndex": 2,
          }
        }
        value="|
|
|
|"
      />
      <div
        style={
          Object {
            "position": "relative",
          }
        }
      >
        <textarea
          cols={40}
          maxLength={251}
          onChange={[Function]}
          onClick={[Function]}
          onKeyUp={[Function]}
          onPaste={[Function]}
          style={
            Object {
              "background": "transparent",
              "border": "none",
              "color": "#1976d2",
              "fontSize": "1.2rem",
              "margin": 0,
              "outline": "none",
              "padding": 0,
              "position": "relative",
              "resize": "none",
              "zIndex": 2,
            }
          }
          value="                                        
                                        
                                        
                                        
                                        
                                        "
        />
        <textarea
          cols={40}
          maxLength={251}
          readOnly={true}
          style={
            Object {
              "background": "transparent",
              "border": "none",
              "color": "#fff",
              "fontSize": "1.2rem",
              "left": "0",
              "margin": 0,
              "outline": "none",
              "padding": 0,
              "position": "absolute",
              "resize": "none",
              "translate": "0",
              "zIndex": 1,
            }
          }
          value="----------------------------------------
----------------------------------------
----------------------------------------
----------------------------------------
----------------------------------------
----------------------------------------"
        />
      </div>
    </div>
  </div>
</div>
`;
